### ISSUE TITLE
Getting 'CredentialsError: Missing credentials in config' when using assume role

### ISSUE THREAD

**Issue by duxing** [aws/aws-sdk-js#1955]
version: `aws-sdk@2.205.0`
OS: `macOS Sierra 10.12.6`

Issue:
I've been using access key to run aws cli for a while and today I'm asked to switch to using assume role. setting up is simple (attached) and I've verified assumed role working with aws cli. then I switch back to working on my nodejs application, and received an error I haven't seen before. 

Error:
```
{ [CredentialsError: Missing credentials in config]
  message: 'Missing credentials in config',
  code: 'CredentialsError',
  errno: 'EHOSTUNREACH',
  syscall: 'connect',
  address: '<my_ip>',
  port: <my_port>,
  time: Thu Mar 01 2018 14:50:33 GMT-0800 (PST),
  originalError: 
   { message: 'Could not load credentials from any providers',
     code: 'CredentialsError',
     errno: 'EHOSTUNREACH',
     syscall: 'connect',
     address: '<my_ip>',
     port: <my_port>,
     time: Thu Mar 01 2018 14:50:33 GMT-0800 (PST),
     originalError: 
      { code: 'EHOSTUNREACH',
        errno: 'EHOSTUNREACH',
        syscall: 'connect',
        address: '<my_ip>',
        port: 80,
        message: 'connect EHOSTUNREACH <my_ip> - Local (<my_local_ip>)' } } }
```

Related issue:
#1483 

My set up:
`~/.aws/config`
```
[profile default]
region = <my_region>
role_arn = <arn_for_assume_role>
source_profile = <my_profile_name>
```

`~/.aws/credentials`
```
[my_profile_name]
aws_access_key_id = <my_access_key>
aws_secret_access_key = <my_secret>
```

`~/.bash_profile`
```
export AWS_SDK_LOAD_CONFIG=1
export AWS_SHARED_CREDENTIALS_FILE=$HOME/.aws/credentials
export AWS_CONFIG_FILE=$HOME/.aws/config
```

Reading #1483 helped me understanding this issue by a lot but still not enough to get it fixed:
 - I've logged `process.env` at runtime and it does include all env variables I exported from `bash_profile`
 - I've logged `aws.config` (`var aws = require('aws-sdk');`):
    - `aws.config.credentials` is `null` at runtime
    - `aws.config.region` is correct

I can get around this issue quickly by change the profile name to `default` in my `~/.aws/credentials` to
```
[default]
aws_access_key_id = <my_access_key>
aws_secret_access_key = <my_secret>
```

With that change, I can get this as `aws.config.credentials` at runtime:
```
SharedIniFileCredentials {
     expired: false,
     expireTime: null,
     accessKeyId: '<my_access_key>',
     sessionToken: undefined,
     filename: undefined,
     profile: 'default',
     disableAssumeRole: true,
     preferStaticCredentials: false }
```

I noticed there is an argument for `SharedIniFileCredentials` called `disableAssumeRole` and it's `true` when i used key access. I'm wondering if you can show me what did I miss here in order to access AWS via node js sdk with assume role at run-time.

Many thanks in advance!

The test code I have:
```
var aws = require('aws-sdk');

aws.config.update({
  region : '<my_region>'
});

var s3 = new aws.S3();
// Note: the bucket and key used here have been verified by accessing via aws cli, using either assume-role or key access
s3.getObject({
  Bucket : '<my_bucket>',
  Key : '<key_to_my_file>'
}).promise().then(function (s3obj){
  console.log('successful');
}).catch(function (error) {
  console.error(error);
});

```

**Comment by chrisradek** [comment]
@duxing 
I think the reason you aren't able to assume a role using the default credentials is because your `roleArn` is supplied in the `~/.aws/config` file instead of the `~/.aws/credentials` file.

If you set the environment variable `AWS_SDK_LOAD_CONFIG` to a truthy value (like 1), then the SDK will load data from the config file as well. This is currently opt-in because support for the config file was added well after the credentials file, and there was a backwards-incompatibility risk otherwise.

Let me know if this doesn't solve the issue for you! I admittedly couldn't find this in our docs so I'm adding a note to update them as well.

**Comment by duxing** [comment]
hi @chrisradek 

I have done that already. see the part I posted about my `.bash_profile` file.
I've also verified this at runtime by logging `process.env` and verified it was set to 1 as an available env variable to the process

**Comment by chrisradek** [comment]
Ah, sorry I missed that. Your configuration looks correct, can you try instantiating the `SharedIniFileCredentials` directly and see if you can assume a role with that?

You should be able to just instantiate it without options if your environment variables are set.
```javascript
var aws = require('aws-sdk');
aws.config.update({
  region: 'you_region'
  credentials: new aws.SharedIniFileCredentials()
});
```
I have to double-check that we will actually assume role by default. If I'm remembering correctly, we didn't want that behavior without a major version bump, but I'll confirm.

**Comment by duxing** [comment]
thanks for the quick response.
I just checked that and got the same error message.
I did see `aws.config.credentials` being forced to `SharedIniFileCredentials` and the `disableAssumeRole` was `false` in this case.

so I guess it's safe to assume this flag is not the cause of the issue.

I'm not familiar with the source code, but setting up a small project should be trivial (all information needed has been listed). do you think you can take some time to debug this when available?

**Comment by chrisradek** [comment]
Thanks @duxing 

The other thing you could try is copying the settings from your config file to your credentials file. That shouldn't be necessary but it might unblock you for now and help me narrow down the problem.

**Comment by duxing** [comment]
still no luck :(
with my set up it only works when I have an access-key-based `default` profile under `~/.aws/credentials`

do you know why the setting for telling aws-sdk to read from `~/.aws/credentials` wasn't working as expected?

**Comment by duxing** [comment]
I did another test:
 - move the credentials from `~/.aws/credentials` to `~/.aws/config` and name the profile as `default`
 - commented out the few AWS environment variable related lines in `~/.bash_profile` and source it again.

running my sample code and I got successful response.

seems like the sdk is smart enough to figure out the credentials even when they're located in the wrong location and the environment variables are off.

can you point me to the place in source code where credentials are loaded from disk, as well as the part for picking which file to load from? I feel debugging this is the best approach

**Comment by chrisradek** [comment]
The SharedIniFileCredentials class can be found [here](https://github.com/aws/aws-sdk-js/blob/master/lib/credentials/shared_ini_file_credentials.js).

I'm also working on reproducing the error, but please share your findings if you dig into it. I'll do the same.

**Comment by chrisradek** [comment]
@duxing 
It looks like the issue comes down to this line:
https://github.com/aws/aws-sdk-js/blob/0fb21c06f67fdc960882aeffe3a01e9ab174e2c9/lib/shared_ini.js#L54

When reading the `config` file, the `default` profile is stored internally as `profile default`, but that line always retrieves the `default` profile regardless of if the config or credentials file is being read. 

If I used any profile other than the `[profile default]` in the `.aws/config` file to specify the role_arn and source_profile, and specified the source_profile in the `.aws/credentials` file, that works.

I'm marking this as a bug since I don't see why we would want this to be the expected behaviour.

**Comment by chrisradek** [comment]
Ah, checking with the CLI, the `~/.aws/config` file treats the `default` profile as special.
If you take a look at their docs [here](https://docs.aws.amazon.com/cli/latest/userguide/cli-multiple-profiles.html), you'll see that the `default` profile is not prepended with `profile`.

Since we're matching the CLI behaviour, this is actually expected. Can you update your shared config file to change the default profile name?

**Comment by duxing** [comment]
@chrisradek sry I was afk for the past 30mins. thanks for digging into this.
I didn't even notice the `profile default` thing while writing this issue down. 

after correcting that to `default`, the configuration I had started working. 
thanks!

**Comment by bahtou** [comment]
Having a similar issue as @duxing 
version: `aws-sdk@2..263.1`
OS: `macOS Sierra 10.13.4`

Sample code:
```
var AWS = require('aws-sdk');
AWS.config.update({
  region: 'us-east-1'
});

var sts = new AWS.STS();
sts.assumeRole({
  RoleArn: <arn-role>
  RoleSessionName: <session-name>
}, function(err, data) {
  if (err) { // an error occurred
    console.log('Cannot assume role');
    console.log(err, err.stack);
  } else { // successful response
    AWS.config.update({
      accessKeyId: data.Credentials.AccessKeyId,
      secretAccessKey: data.Credentials.SecretAccessKey,
      sessionToken: data.Credentials.SessionToken
    });
  }
});
```
I followed the steps outlined in this thread and still unable to successfully assumeRole.

`.aws/config`
```
[default]
region=us-east-1
role_arn=<arn-role>
source_profile=default
aws_access_key_id=<access-key>
aws_secret_access_key=<access-secret>
```

I commented out everything in the `credentials` file and also in the `.bash_profile`. 
```
SharedIniFileCredentials {
  expired: false,
  expireTime: null,
  accessKeyId: undefined,
  sessionToken: undefined,
  filename: undefined,
  profile: 'default',
  disableAssumeRole: false,
  preferStaticCredentials: false }
```

**Comment by shyamkumar123** [comment]
hi CredentialsError: Missing credentials in config
    at ClientRequest.<anonymous> (C:\Users\Zinios Dev\Downloads\image-rekognition\node_modules\aws-sdk\lib\http\node.js:83:34)    at Object.onceWrapper (events.js:313:30)    at emitNone (events.js:106:13)
    at ClientRequest.emit (events.js:208:7)
    at Socket.emitTimeout (_http_client.js:706:34)
    at Object.onceWrapper (events.js:313:30)
    at emitNone (events.js:106:13)
    at Socket.emit (events.js:208:7)
    at Socket._onTimeout (net.js:410:8)
    at ontimeout (timers.js:498:11)
  message: 'Missing credentials in config',
  code: 'CredentialsError',
  time: 2018-08-06T18:05:45.323Z,
  retryable: true,
  originalError:
   { message: 'Could not load credentials from any providers',
     code: 'CredentialsError',
     time: 2018-08-06T18:05:45.323Z,
     retryable: true,
     originalError:
      { message: 'Connection timed out after 1000ms',
        code: 'TimeoutError',
        time: 2018-08-06T18:05:45.323Z,
        retryable: true } } } 'CredentialsError: Missing credentials in config\n    at ClientRequest.<anonymous> (C:\\Users\\Zinios Dev\\Downloads\\image-rekognition\\node_modules\\aws-sdk\\lib\\http\\node.js:83:34)\n    at Object.onceWrapper (events.js:313:30)\n    at emitNone (events.js:106:13)\n    at ClientRequest.emit (events.js:208:7)\n    at Socket.emitTimeout (_http_client.js:706:34)\n    at Object.onceWrapper (events.js:313:30)\n    at emitNone (events.js:106:13)\n    at Socket.emit (events.js:208:7)\n    at Socket._onTimeout (net.js:410:8)\n    at ontimeout (timers.js:498:11)'
{ CredentialsError: Missing credentials in config
    at ClientRequest.<anonymous> (C:\Users\Zinios Dev\Downloads\image-rekognition\node_modules\aws-sdk\lib\http\node.js:83:34)
    at Object.onceWrapper (events.js:313:30)
    at emitNone (events.js:106:13)
    at ClientRequest.emit (events.js:208:7)
    at Socket.emitTimeout (_http_client.js:706:34)
    at Object.onceWrapper (events.js:313:30)
    at emitNone (events.js:106:13)
    at Socket.emit (events.js:208:7)
    at Socket._onTimeout (net.js:410:8)
    at ontimeout (timers.js:498:11)
  message: 'Missing credentials in config',
  code: 'CredentialsError',
  time: 2018-08-06T18:05:45.332Z,
  retryable: true,
  originalError:
   { message: 'Could not load credentials from any providers',
     code: 'CredentialsError',
     time: 2018-08-06T18:05:45.332Z,
     retryable: true,
     originalError:
      { message: 'Missing credentials in config',
        code: 'CredentialsError',
        time: 2018-08-06T18:05:45.323Z,
        retryable: true,
        originalError: [Object] } } } 'CredentialsError: Missing credentials in config\n    at ClientRequest.<anonymous>
(C:\\Users\\Zinios Dev\\Downloads\\image-rekognition\\node_modules\\aws-sdk\\lib\\http\\node.js:83:34)\n    at Object.onceWrapper (events.js:313:30)\n    at emitNone (events.js:106:13)\n    at ClientRequest.emit (events.js:208:7)\n    at Socket.emitTimeout (_http_client.js:706:34)\n    at Object.onceWrapper (events.js:313:30)\n    at emitNone (events.js:106:13)\n    at Socket.emit (events.js:208:7)\n    at Socket._onTimeout (net.js:410:8)\n    at ontimeout (timers.js:498:11)'


how to solve above issue ,please help me

**Comment by chemitaxis** [comment]
Same error here... All working fine during months, today Sentry has logged this error in my EC2 instance... 

// timeout support
    stream.setTimeout(httpOptions.timeout || 0, function() {
      if (stream.didCallback) return; stream.didCallback = true;
      var msg = 'Connection timed out after ' + httpOptions.timeout + 'ms';
      errCallback(AWS.util.error(new Error(msg), {code: 'TimeoutError'}));
      stream.abort();
    });

**Comment by ted-intradenver** [comment]
Making sure the AWS CLI is running and the default user is configured helps. (note: how to Mac OSX brew install here: https://www.code2bits.com/how-to-install-awscli-on-macos-using-homebrew/) Then (https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html).

**Comment by brendadev** [comment]
Just wanted to add here that I recently encountered this `Missing credentials in config` error, while using `aws-sdk`. I was sure this was working before, so I tested the `aws cli` and saved a file into S3 without issue. At that point the only difference was the use of the `aws-sdk`, so I upgrade and surprise, `Missing credentials in config` was gone and I was able to save a file in S3 using `aws-sdk`.
